File: go/test/log_drain_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/vercel-sdk/go"
	"github.com/voxgig-sdk/vercel-sdk/go/core"

	vs "github.com/voxgig-sdk/vercel-sdk/go/utility/struct"
)

func TestLogDrainEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.LogDrain(nil)
		if ent == nil {
			t.Fatal("expected non-nil LogDrainEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"log_drain": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.LogDrain(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.LogDrain(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := log_drainBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "log_drain." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set VERCEL_TEST_LOG_DRAIN_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		logDrainRef01Ent := client.LogDrain(nil)
		logDrainRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "log_drain"}), "log_drain_ref01"))

		logDrainRef01DataResult, err := logDrainRef01Ent.Create(logDrainRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		logDrainRef01Data = core.ToMapAny(entityData(logDrainRef01DataResult))
		if logDrainRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if logDrainRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		logDrainRef01Match := map[string]any{}

		logDrainRef01ListResult, err := logDrainRef01Ent.List(logDrainRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		logDrainRef01List, logDrainRef01ListOk := logDrainRef01ListResult.([]any)
		if !logDrainRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", logDrainRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(logDrainRef01List), map[string]any{"id": logDrainRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// LOAD
		logDrainRef01MatchDt0 := map[string]any{
			"id": logDrainRef01Data["id"],
		}
		logDrainRef01DataDt0Loaded, err := logDrainRef01Ent.Load(logDrainRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		logDrainRef01DataDt0LoadResult := core.ToMapAny(entityData(logDrainRef01DataDt0Loaded))
		if logDrainRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if logDrainRef01DataDt0LoadResult["id"] != logDrainRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		logDrainRef01MatchRm0 := map[string]any{
			"id": logDrainRef01Data["id"],
		}
		_, err = logDrainRef01Ent.Remove(logDrainRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		logDrainRef01MatchRt0 := map[string]any{}

		logDrainRef01ListRt0Result, err := logDrainRef01Ent.List(logDrainRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		logDrainRef01ListRt0, logDrainRef01ListRt0Ok := logDrainRef01ListRt0Result.([]any)
		if !logDrainRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", logDrainRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(logDrainRef01ListRt0), map[string]any{"id": logDrainRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func log_drainBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "log_drain", "LogDrainTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read log_drain test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse log_drain test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"log_drain01", "log_drain02", "log_drain03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("VERCEL_TEST_LOG_DRAIN_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"VERCEL_TEST_LOG_DRAIN_ENTID": idmap,
		"VERCEL_TEST_LIVE":      "FALSE",
		"VERCEL_TEST_EXPLAIN":   "FALSE",
		"VERCEL_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["VERCEL_TEST_LOG_DRAIN_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["VERCEL_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["VERCEL_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewVercelSDK(core.ToMapAny(mergedOpts))
	}

	live := env["VERCEL_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["VERCEL_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/log_drain_entity_test.lua
-- LogDrain entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("vercel_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("LogDrainEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:LogDrain(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["log_drain"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:LogDrain(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:LogDrain(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = log_drain_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "log_drain." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set VERCEL_TEST_LOG_DRAIN_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local log_drain_ref01_ent = client:LogDrain(nil)
    local log_drain_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.log_drain"), "log_drain_ref01"))

    local log_drain_ref01_data_result, err = log_drain_ref01_ent:create(log_drain_ref01_data, nil)
    assert.is_nil(err)
    log_drain_ref01_data = helpers.to_map(type(log_drain_ref01_data_result) == 'table' and log_drain_ref01_data_result.data_get and log_drain_ref01_data_result:data_get() or log_drain_ref01_data_result)
    assert.is_not_nil(log_drain_ref01_data)
    assert.is_not_nil(log_drain_ref01_data["id"])

    -- LIST
    local log_drain_ref01_match = {}

    local log_drain_ref01_list_result, err = log_drain_ref01_ent:list(log_drain_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(log_drain_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(log_drain_ref01_list_result),
      { id = log_drain_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- LOAD
    local log_drain_ref01_match_dt0 = {
      id = log_drain_ref01_data["id"],
    }
    local log_drain_ref01_data_dt0_loaded, err = log_drain_ref01_ent:load(log_drain_ref01_match_dt0, nil)
    assert.is_nil(err)
    local log_drain_ref01_data_dt0_load_result = helpers.to_map(type(log_drain_ref01_data_dt0_loaded) == 'table' and log_drain_ref01_data_dt0_loaded.data_get and log_drain_ref01_data_dt0_loaded:data_get() or log_drain_ref01_data_dt0_loaded)
    assert.is_not_nil(log_drain_ref01_data_dt0_load_result)
    assert.are.equal(log_drain_ref01_data_dt0_load_result["id"], log_drain_ref01_data["id"])

    -- REMOVE
    local log_drain_ref01_match_rm0 = {
      id = log_drain_ref01_data["id"],
    }
    local _, err = log_drain_ref01_ent:remove(log_drain_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local log_drain_ref01_match_rt0 = {}

    local log_drain_ref01_list_rt0_result, err = log_drain_ref01_ent:list(log_drain_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(log_drain_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(log_drain_ref01_list_rt0_result),
      { id = log_drain_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function log_drain_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/log_drain/LogDrainTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read log_drain test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "log_drain01", "log_drain02", "log_drain03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("VERCEL_TEST_LOG_DRAIN_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["VERCEL_TEST_LOG_DRAIN_ENTID"] = idmap,
    ["VERCEL_TEST_LIVE"] = "FALSE",
    ["VERCEL_TEST_EXPLAIN"] = "FALSE",
    ["VERCEL_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["VERCEL_TEST_LOG_DRAIN_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["VERCEL_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["VERCEL_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["VERCEL_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["VERCEL_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
